zaza/utilities/openstack.py

```diff
@@ -1269,3 +1285,179 @@ def get_overcloud_auth():
+        "Resource in {} state, waiting for {}" .format(resource_stat,
+                                                       expected_stat,))
+
+
+@tenacity.retry(wait=tenacity.wait_exponential(multiplier=1, max=60),
+                reraise=True, stop=tenacity.stop_after_attempt(2))
+def resource_removed(resource, resource_id, msg="resource"):
+    """Wait for an openstack resource to no longer be present
+
+    :param resource: pointer to os resource type, ex: heat_client.stacks
+    :type resource: str
+    :param resource_id: unique id for the openstack resource
+    :type resource_id: str
+    :param msg: text to identify purpose in logging
+    :type msy: str
+    :raises: AssertionError
+    """
+    matching = [r for r in resource.list() if r.id == resource_id]
+    logging.debug("Resource {} still present".format(resource_id))
+    assert len(matching) == 0, "Resource {} still present".format(resource_id)
+
+
+def delete_resource(resource, resource_id, msg="resource"):
+    """Delete an openstack resource, such as one instance, keypair,
+    image, volume, stack, etc., and confirm deletion within max wait time.
+
+    :param resource: pointer to os resource type, ex:glance_client.images
+    :type resource: str
+    :param resource_id: unique name or id for the openstack resource
+    :type resource_id: str
+    :param msg: text to identify purpose in logging
+    :type msg: str
+    """
+    logging.debug('Deleting OpenStack resource '
+                  '{} ({})'.format(resource_id, msg))
+    resource.delete(resource_id)
+    resource_removed(resource, resource_id, msg)
+
+
+def delete_image(glance, img_id):
+    """Delete the given image from glance
+
+    :param glance: Authenticated glanceclient
+    :type glance: glanceclient.Client
+    :param img_id: unique name or id for the openstack resource
+    :type img_id: str
+    """
+    delete_resource(glance.images, img_id, msg="glance image")
+
+
+def upload_image_to_glance(glance, local_path, image_name):
+    """Upload the given image to glance and apply the given label
+
+    :param glance: Authenticated glanceclient
+    :type glance: glanceclient.Client
+    :param local_path: Path to local image
+    :type local_path: str
+    :param image_name: The label to give the image in glance
+    :type image_name: str
+    """
+    # Create glance image
+    image = glance.images.create(
+        name=image_name,
+        disk_format='qcow2',
+        visibility='public',
+        container_format='bare')
+    glance.images.upload(image.id, open(local_path, 'rb'))
+
+    resource_reaches_status(
+        glance.images,
+        image.id,
+        expected_stat='active',
+        msg='Image status wait')
+
+    return image
+
+
+def create_image(glance, image_url, image_name, image_cache_dir='tests'):
+    """Download the latest cirros image and upload it to glance,
+    validate and return a resource pointer.
+
+    :param glance: pointer to authenticated glance connection
```

Comment:
For consistency let's stick with:
 :param glance: Authenticated glanceclient

Reply:
Done